File: samples/VB.NET/ProcessScannedDocumentsWithBarcodeSeparators/ProcessScannedDocumentsWithBarcodeSeparators/Form1.vb
Imports Dynamsoft.DotNet.TWAIN
Imports Dynamsoft.DotNet.TWAIN.Enums.Barcode

Public Class Form1

    Dim m_barcodeformat As String

    Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        Me.Icon = New Icon(GetType(Form), "wfc.ico")
        Me.dynamicDotNetTwain1.Visible = False
        Me.cmbBarcodeFormat.DropDownStyle = ComboBoxStyle.DropDownList
        Me.cmbBarcodeFormat.DataSource = [Enum].GetValues(GetType(BarcodeFormat))
        Me.cmbBarcodeFormat.SelectedIndex = 0

        InitPicMode()

        Dim strDllPath As String
        Dim m_strCurrentDirectory As String
        m_strCurrentDirectory = Application.ExecutablePath
        Dim pos As Integer
        pos = m_strCurrentDirectory.LastIndexOf("\Samples\")
        If (pos <> -1) Then
            m_strCurrentDirectory = m_strCurrentDirectory.Substring(0, m_strCurrentDirectory.IndexOf("\", pos)) + "\"
            strDllPath = m_strCurrentDirectory + "Redistributable\BarcodeResources\"
        Else
            pos = m_strCurrentDirectory.LastIndexOf("\")
            m_strCurrentDirectory = m_strCurrentDirectory.Substring(0, m_strCurrentDirectory.IndexOf("\", pos)) + "\"
            strDllPath = m_strCurrentDirectory
        End If
        Me.dynamicDotNetTwain1.BarcodeDllPath = strDllPath

        dynamicDotNetTwain1.IfShowCancelDialogWhenBarcodeOrOCR = True
        dynamicDotNetTwain1.MaxImagesInBuffer = 64
        dynamicDotNetTwain1.ScanInNewProcess = True

        Dim toolTip1, toolTip2 As New ToolTip()
        toolTip1.AutoPopDelay = 5000
        toolTip1.InitialDelay = 1000
        toolTip1.ReshowDelay = 500
        toolTip1.ShowAlways = True
        toolTip1.SetToolTip(Me.picFAQMode1, "All pages in all source files are considered " & vbCrLf & "as one uninterrupted sequence of pages. " & vbCrLf & "Destination files are arranged so that " & vbCrLf & "the first page always contains a barcode. ")

        toolTip2.AutoPopDelay = 5000
        toolTip2.InitialDelay = 1000
        toolTip2.ReshowDelay = 500
        toolTip2.ShowAlways = True
        toolTip2.SetToolTip(Me.picFAQMode2, "Each page contains a barcode and the pages " & vbCrLf & "where barcodes coincide are detected and " & vbCrLf & "got merged to a PDF file.")


    End Sub

    Private Sub radMode1_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles radMode1.CheckedChanged
        InitPicMode()
    End Sub

    Private Sub radMode2_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles radMode2.CheckedChanged
        InitPicMode()
    End Sub

    Private Sub InitPicMode()
        If (radMode1.Checked = True) Then
            Me.picMode1.Image = Global.PSDWBS.My.Resources.Resources.Mode1_Selected
            Me.picMode2.Image = Global.PSDWBS.My.Resources.Resources.Mode2
        Else
            Me.picMode1.Image = Global.PSDWBS.My.Resources.Resources.Mode2
            Me.picMode2.Image = Global.PSDWBS.My.Resources.Resources.Mode2_Selected
        End If
    End Sub

    Dim m_listBarcodeType As List(Of String)

    Private Sub SaveFileByBarcodeText()
        m_listBarcodeType = New List(Of String)
        Dim listImageIndex As New List(Of IndexList)
        Dim listIndex As New IndexList
        listImageIndex.Add(listIndex)  'use to save no barcode files

        Dim i As Integer

        For i = 0 To Me.dynamicDotNetTwain1.HowManyImagesInBuffer - 1
            Dim aryResult As Barcode.Result()
            Dim format As BarcodeFormat
            format = CType(System.Enum.Parse(GetType(BarcodeFormat), Val(cmbBarcodeFormat.SelectedValue)), BarcodeFormat)
            aryResult = Me.dynamicDotNetTwain1.ReadBarcode(Convert.ToInt16(i), format) 'Please update the barcode format to yours


            If (aryResult Is Nothing Or aryResult.Length = 0) Then
                'If no barcode found on the current image, add it to the image list for saving
                UpdateDateList(0, i, listImageIndex)
            Else 'If a barcode is found, restart the list
                Dim strBarcodeText As String
                strBarcodeText = aryResult(0).Text
                Dim iPosition As Integer
                iPosition = 0
                If (IfExistBarcodeText(strBarcodeText, iPosition) = True) Then
                    UpdateDateList(iPosition, i, listImageIndex)
                Else
                    m_listBarcodeType.Add(strBarcodeText)
                    AddDateList(i, listImageIndex)
                End If
            End If
        Next

        SaveImages(listImageIndex)

    End Sub

    Private Sub AddDateList(ByVal index As Integer, ByRef listImageIndex As List(Of IndexList))
        Dim listIndex As New IndexList
        listIndex.Add(index)
        listImageIndex.Add(listIndex)
    End Sub


    Private Sub UpdateDateList(ByVal iPosition As Integer, ByVal index As Integer, ByRef listImageIndex As List(Of IndexList))
        Dim listIndex As New IndexList
        listIndex = listImageIndex(iPosition)
        listIndex.Add(index)
        listImageIndex(iPosition) = listIndex
    End Sub


    Private Function IfExistBarcodeText(ByVal strBarcodeText As String, ByRef iPosition As Integer) As Boolean
        iPosition = 0
        Dim bRet As Boolean
        bRet = False
        Dim strTemp As String
        strTemp = ""
        Dim i As Integer

        For i = 0 To m_listBarcodeType.Count - 1
            strTemp = m_listBarcodeType(i)
            If (strBarcodeText.Trim().ToLower().CompareTo(strTemp.Trim().ToLower()) = 0) Then
                iPosition = i + 1
                bRet = True
                Exit For
            End If
        Next
        Return bRet
    End Function


    Private Sub SaveImages(ByVal listImageIndex As List(Of IndexList))
        Dim index As Integer
        index = 0
        Dim objFolderBrowserDialog As New FolderBrowserDialog

        If (objFolderBrowserDialog.ShowDialog() = DialogResult.OK) Then
            Dim list As IndexList
            For Each list In listImageIndex
                If (list.Count <> 0) Then
                    Me.dynamicDotNetTwain1.SaveAsMultiPagePDF(objFolderBrowserDialog.SelectedPath.Trim() + "\" + index.ToString() + ".pdf", list)
                End If
                index = index + 1
            Next
        End If

        MessageBox.Show(Me.dynamicDotNetTwain1.ErrorString)
    End Sub


    Private Sub SaveFileByBegainWithBarcode()
        Dim listImageIndex As New List(Of IndexList)
        Dim listIndex As IndexList
        Dim i As Integer
        For i = 0 To Me.dynamicDotNetTwain1.HowManyImagesInBuffer - 1
            If listIndex Is Nothing Then
                listIndex = New IndexList()
            End If

            Dim aryResult As Barcode.Result()
            Dim format As BarcodeFormat
            format = CType(System.Enum.Parse(GetType(BarcodeFormat), Val(cmbBarcodeFormat.SelectedValue)), BarcodeFormat)
            aryResult = Me.dynamicDotNetTwain1.ReadBarcode(Convert.ToInt16(i), format) 'Please update the barcode format to yours


            If (aryResult Is Nothing Or aryResult.Length = 0) Then
                listIndex.Add(i) 'If no barcode found on the current image, add it to the image list for saving
            Else
                If (Not listIndex Is Nothing And listIndex.Count > 0) Then
                    listImageIndex.Add(listIndex)
                    listIndex = Nothing
                End If

                'If a barcode is found, restart the list
                If listIndex Is Nothing Then
                    listIndex = New IndexList()
                End If
                listIndex.Add(i)
            End If

        Next

        If (Not listIndex Is Nothing) Then
            listImageIndex.Add(listIndex)  'save a last set of data
            listIndex = Nothing
        End If

        SaveImages(listImageIndex)
    End Sub


    Private Sub btnScan_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnScan.Click
        Me.dynamicDotNetTwain1.SelectSource()
        Me.dynamicDotNetTwain1.OpenSource()
        Me.dynamicDotNetTwain1.IfAutoFeed = True
        Me.dynamicDotNetTwain1.IfFeederEnabled = True
        Me.dynamicDotNetTwain1.IfShowUI = False
        Me.dynamicDotNetTwain1.AcquireImage()
    End Sub

    Private Sub btnRemoveAllImage_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnRemoveAllImage.Click
        Me.dynamicDotNetTwain1.RemoveAllImages()
        Me.dynamicDotNetTwain1.Visible = False
    End Sub

    Private Sub btnRemoveSelectedImages_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnRemoveSelectedImages.Click
        Me.dynamicDotNetTwain1.RemoveImages(Me.dynamicDotNetTwain1.CurrentSelectedImageIndicesInBuffer)
        If Me.dynamicDotNetTwain1.HowManyImagesInBuffer = 0 Then
            Me.dynamicDotNetTwain1.Visible = False
        End If
    End Sub

    Private Sub btnSave_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSave.Click
        If (Me.dynamicDotNetTwain1.HowManyImagesInBuffer > 0) Then
            If (Me.radMode1.Checked = True) Then
                SaveFileByBegainWithBarcode()
            Else
                SaveFileByBarcodeText()
            End If
        End If
    End Sub

    Private Sub dynamicDotNetTwain1_OnPostAllTransfers() Handles dynamicDotNetTwain1.OnPostAllTransfers
        If (Me.dynamicDotNetTwain1.Visible = False) Then
            Me.dynamicDotNetTwain1.Visible = True
        End If
    End Sub
End Class

File: samples/C#/ProcessScannedDocumentsWithBarcodeSeparators/ProcessScannedDocumentsWithBarcodeSeparators/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using Dynamsoft.DotNet.TWAIN;
using Dynamsoft.DotNet.TWAIN.Enums.Barcode;
using Dynamsoft.DotNet.TWAIN.Barcode;

namespace AddBarcodeDemo
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
            Initialization();       
        }

        protected void Initialization()
        {
            this.Icon = new Icon(typeof(Form), "wfc.ico");
            this.dynamicDotNetTwain1.Visible = false;
            this.cmbBarcodeFormat.DropDownStyle = ComboBoxStyle.DropDownList;
            this.cmbBarcodeFormat.DataSource = Enum.GetValues(typeof(Dynamsoft.DotNet.TWAIN.Enums.Barcode.BarcodeFormat));
            this.cmbBarcodeFormat.SelectedIndex = 0;

            InitPicMode(); 
            
            string strDllFolder = Application.ExecutablePath;
            strDllFolder = strDllFolder.Replace("/", "\\");
            int pos = strDllFolder.LastIndexOf("\\Samples\\");
            if (pos != -1)
            {
                strDllFolder = strDllFolder.Substring(0, strDllFolder.IndexOf(@"\", pos)) + @"\Redistributable\BarcodeResources\";
            }
            else
            {
                pos = strDllFolder.LastIndexOf("\\");
                strDllFolder = strDllFolder.Substring(0, strDllFolder.IndexOf(@"\", pos)) + @"\";
             }

            dynamicDotNetTwain1.BarcodeDllPath = strDllFolder;

            dynamicDotNetTwain1.IfShowCancelDialogWhenBarcodeOrOCR = true;
            dynamicDotNetTwain1.MaxImagesInBuffer = 64;
            dynamicDotNetTwain1.ScanInNewProcess = true;

            ToolTip toolTip1 = new ToolTip();
            toolTip1.AutoPopDelay = 5000;
            toolTip1.InitialDelay = 1000;
            toolTip1.ReshowDelay = 500;
            toolTip1.ShowAlways = true;
            toolTip1.SetToolTip(this.picFAQMode1, "All pages in all source files are considered \r\nas one uninterrupted sequence of pages. \r\nDestination files are arranged so that \r\nthe first page always contains a barcode. ");
 
            ToolTip toolTip2 = new ToolTip();
            toolTip2.AutoPopDelay = 5000;
            toolTip2.InitialDelay = 1000;
            toolTip2.ReshowDelay = 500;
            toolTip2.ShowAlways = true;
            toolTip2.SetToolTip(this.picFAQMode2, "Each page contains a barcode and the pages \r\nwhere barcodes coincide are detected and \r\ngot merged to a PDF file.");
        }

        private void radMode1_CheckedChanged(object sender, EventArgs e)
        {
            InitPicMode();
        }

        private void radMode2_CheckedChanged(object sender, EventArgs e)
        {
            InitPicMode();
        }

        private void InitPicMode()
        {
            if (radMode1.Checked == true)
            {
                this.picMode1.Image = global::PSDWBS.Properties.Resources.Mode1_Selected;
                this.picMode2.Image = global::PSDWBS.Properties.Resources.Mode2;
            }
            else
            {
                this.picMode1.Image = global::PSDWBS.Properties.Resources.Mode1;
                this.picMode2.Image = global::PSDWBS.Properties.Resources.Mode2_Selected;
            }
        }

        List<string> m_listBarcodeType = null;
        private void SaveFileByBarcodeText()
        {
            m_listBarcodeType = new List<string>();
            List<IndexList> listImageIndex = new List<IndexList>();
            IndexList listIndex = new IndexList();
            listImageIndex.Add(listIndex); //use to save no barcode files

            for (int i = 0; i < this.dynamicDotNetTwain1.HowManyImagesInBuffer; i++)
            {

                Result[] aryResult = this.dynamicDotNetTwain1.ReadBarcode((short)i, (BarcodeFormat)cmbBarcodeFormat.SelectedValue);//Please update the barcode format to yours


                if (null == aryResult || aryResult.Length == 0)
                {
                    //If no barcode found on the current image, add it to the image list for saving
                    UpdateDateList(0, i, ref listImageIndex);
                }
                else //If a barcode is found, restart the list
                {
                    string strBarcodeText = aryResult[0].Text;
                    int iPosition = 0;
                    if (IfExistBarcodeText(strBarcodeText, out iPosition))
                    {
                        UpdateDateList(iPosition, i, ref listImageIndex);
                    }
                    else
                    {
                        m_listBarcodeType.Add(strBarcodeText);

                        AddDateList(i, ref listImageIndex);
                    }
                }
            }
            SaveImages(listImageIndex);
        }

        private void AddDateList(int index, ref List<IndexList> listImageIndex)
        {
            IndexList listIndex = new IndexList();
            listIndex.Add(index);
            listImageIndex.Add(listIndex);
        }

        private void UpdateDateList(int iPosition, int index, ref List<IndexList> listImageIndex)
        {
            IndexList listIndex = new IndexList();
            listIndex = listImageIndex[iPosition];
            listIndex.Add(index);
            listImageIndex[iPosition] = listIndex;
        }

        private bool IfExistBarcodeText(string strBarcodeText, out int iPosition)
        {
            iPosition = 0;
            bool bRet = false;
            string strTemp = "";
            int i = 0;
            for (i = 0; i < m_listBarcodeType.Count; i++)
            {
                strTemp = m_listBarcodeType[i];
                if (strBarcodeText.Trim().ToLower().CompareTo(strTemp.Trim().ToLower()) == 0)
                {
                    iPosition = i + 1;
                    bRet = true;
                    break;
                }
            }
            return bRet;
        }

        private void SaveImages(List<IndexList> listImageIndex)
        {
            int index = 0;
            FolderBrowserDialog objFolderBrowserDialog = new FolderBrowserDialog();
            if (objFolderBrowserDialog.ShowDialog() == DialogResult.OK)
            {
                foreach (IndexList list in listImageIndex)
                {
                    if (list.Count != 0)
                    {
                        this.dynamicDotNetTwain1.SaveAsMultiPagePDF(objFolderBrowserDialog.SelectedPath.Trim() + "\\" + index + ".pdf", list);
                    }
                    index++;
                }
            }

            MessageBox.Show(this.dynamicDotNetTwain1.ErrorString);
        }


        private void SaveFileByBegainWithBarcode()
        {
            List<IndexList> listImageIndex = new List<IndexList>();
            IndexList listIndex = null;

            for (int i = 0; i < this.dynamicDotNetTwain1.HowManyImagesInBuffer; i++)
            {
                if (null == listIndex)
                    listIndex = new IndexList();

                Result[] aryResult = this.dynamicDotNetTwain1.ReadBarcode((short)i, (BarcodeFormat)cmbBarcodeFormat.SelectedValue);//Please update the barcode format to yours


                if (null == aryResult || aryResult.Length == 0)
                {
                    listIndex.Add(i); //If no barcode found on the current image, add it to the image list for saving
                }
                else
                {
                    if (listIndex != null && listIndex.Count > 0)
                    {
                        listImageIndex.Add(listIndex);
                        listIndex = null;
                    }

                    //If a barcode is found, restart the list
                    if (null == listIndex)
                        listIndex = new IndexList();
                    listIndex.Add(i);
                }
            }

            if (listIndex != null)
            {
                listImageIndex.Add(listIndex);  //save a last set of data
                listIndex = null;
            }

            SaveImages(listImageIndex);
        }

        private void btnScan_Click(object sender, EventArgs e)
        {
            this.dynamicDotNetTwain1.SelectSource();
            this.dynamicDotNetTwain1.OpenSource();
            this.dynamicDotNetTwain1.IfAutoFeed = true;
            this.dynamicDotNetTwain1.IfFeederEnabled = true;
            this.dynamicDotNetTwain1.IfShowUI = false;    
            this.dynamicDotNetTwain1.AcquireImage();
        }

        private void btnRemoveAllImage_Click(object sender, EventArgs e)
        {
            this.dynamicDotNetTwain1.RemoveAllImages();
            this.dynamicDotNetTwain1.Visible = false;
        }

        private void btnRemoveSelectedImages_Click(object sender, EventArgs e)
        {
            this.dynamicDotNetTwain1.RemoveImages(this.dynamicDotNetTwain1.CurrentSelectedImageIndicesInBuffer);
            if(this.dynamicDotNetTwain1.HowManyImagesInBuffer == 0)
                this.dynamicDotNetTwain1.Visible = false;
        }

        private void btnSave_Click(object sender, EventArgs e)
        {
            if (this.dynamicDotNetTwain1.HowManyImagesInBuffer > 0)
            {
                if (this.radMode1.Checked == true)
                    SaveFileByBegainWithBarcode();
                else
                    SaveFileByBarcodeText();
            }
        }

        private void dynamicDotNetTwain1_OnPostAllTransfers()
        {
            if(this.dynamicDotNetTwain1.Visible == false)
                this.dynamicDotNetTwain1.Visible = true;
        }

    }
}
